client/allocwatcher/alloc_watcher.go

```diff
@@ -85,47 +94,103 @@ type Config struct {
 	Logger hclog.Logger
 }
 
-// NewAllocWatcher creates a PrevAllocWatcher appropriate for whether this
-// alloc's previous allocation was local or remote. If this alloc has no
-// previous alloc then a noop implementation is returned.
-func NewAllocWatcher(c Config) PrevAllocWatcher {
-	if c.Alloc.PreviousAllocation == "" {
-		// No previous allocation, use noop transitioner
-		return NoopPrevAlloc{}
+func newMigratorForAlloc(c Config, tg *structs.TaskGroup, watchedAllocID string, m AllocRunnerMeta) PrevAllocMigrator {
+	logger := c.Logger.Named("alloc_migrator")
+	logger = logger.With("alloc_id", c.Alloc.ID)
+	logger = logger.With("previous_alloc", watchedAllocID)
```

Comment:
Nit - any reason why this can't be done fluent style like 
```
    logger := c.Logger.Named("")
                      .With("alloc_id",..)
                      .With("prev_alloc",..)
```